daml-script/daml3/Daml/Script/Questions/Commands.daml

```diff
@@ -74,50 +79,104 @@ instance Applicative Commands where
+createAndExerciseCmdWithCid : forall t c r. (Template t, Choice t c r, HasAgreement t) => t -> c -> Commands (ContractId t, r)
+-- The 'HasAgreement t' constraint prevents this function from being used on interface types.
+createAndExerciseCmdWithCid tplArg choiceArg = internalTypedCreateAndExerciseCmdWithCid tplArg choiceArg False
 
--- | Create a contract and exercise a choice on it in the same transaction.
+-- | Create a contract and exercise a choice on it in the same transaction, returns only the choice result.
 createAndExerciseCmd : forall t c r. (Template t, Choice t c r, HasAgreement t) => t -> c -> Commands r
 -- The 'HasAgreement t' constraint prevents this function from being used on interface types.
-createAndExerciseCmd tplArg choiceArg = fromLedgerValue @r <$> internalCreateAndExerciseCmd (toAnyTemplate tplArg) (toAnyChoice @t choiceArg)
+createAndExerciseCmd tplArg choiceArg = snd <$> createAndExerciseCmdWithCid tplArg choiceArg
+
+-- Main command API, WITH explicit package ids
+-- | Create a contract of the given template, using the exact package ID of the template given - upgrades are disabled.
+createCmdExact : (Template t, HasAgreement t) => t -> Commands (ContractId t)
+-- The 'HasAgreement t' constraint prevents this function from being used on interface types.
+createCmdExact arg = internalTypedCreateCmd arg True
+
+-- | Exercise a choice on the given contract, using the exact package ID of the template given - upgrades are disabled.
+exerciseCmdExact : forall t c r. Choice t c r => ContractId t -> c -> Commands r
+exerciseCmdExact cId arg = internalTypedExerciseCmd cId arg True
+
+-- | Exercise a choice on the contract with the given key, using the exact package ID of the template given - upgrades are disabled.
+exerciseByKeyCmdExact : forall t k c r. (TemplateKey t k, Choice t c r) => k -> c -> Commands r
+exerciseByKeyCmdExact key arg = internalTypedExerciseByKeyCmd @t key arg True
+
+-- | Create a contract and exercise a choice on it in the same transaction, returns the created ContractId, and the choice result.
+-- Uses the exact package ID of the template given - upgrades are disabled.
+createAndExerciseCmdWithCidExact : forall t c r. (Template t, Choice t c r, HasAgreement t) => t -> c -> Commands (ContractId t, r)
+-- The 'HasAgreement t' constraint prevents this function from being used on interface types.
+createAndExerciseCmdWithCidExact tplArg choiceArg = internalTypedCreateAndExerciseCmdWithCid tplArg choiceArg True
+
+-- | Create a contract and exercise a choice on it in the same transaction, returns only the choice result.
+createAndExerciseCmdExact : forall t c r. (Template t, Choice t c r, HasAgreement t) => t -> c -> Commands r
+-- The 'HasAgreement t' constraint prevents this function from being used on interface types.
+createAndExerciseCmdExact tplArg choiceArg = snd <$> createAndExerciseCmdWithCidExact tplArg choiceArg
```

Comment:
Theres a lot of submit functions now, perhaps this can be cleaned up a bit - unsure, most of the file is type signatures...

Reply:
this made me think, since we already have `createAndExercise : ... Update r`, would we also want to add `createAndExerciseExact : ... Update r` at some point? if so, I think we should be consistent and keep `...Cmd` at the end to signal the `: ... Command r` version of an `: ... Update r` function

Reply:
Ah right, so `createAndExerciseExactCmd`. I'll discuss this with Bas to see if this is something we want. I know that upgrade behaviour is slightly different within `Update`, since package ids are already selected, so I'm unsure if this will change things.

Reply:
I've made this change, though we'll likely rethink how this `Exact` version of commands will look, as the ledger api interface for upgrades is changing